### DISCUSSION TITLE
Multiple Signature

### DISCUSSION THREAD

**Discussion by ghoshpushpendu** [MatthiasValvekens/pyHanko#70] [Q&A] [answered]
I have a PDF with more than 30 pages and 1 signer.

I want to put the signature of the signer in every page in same position.

`  pyhanko sign addfields --field 1/445,200,545,250/signature1 --field 2/445,200,545,250/signature2 --field 3/445,200,545,250/signature3 --field 4/445,200,545,250/signature4 --field 5/445,200,545,250/signature5 --field 6/445,200,545,250/signature6 --field 7/445,200,545,250/signature7 --field 8/445,200,545,250/signature8 --field 9/445,200,545,250/signature9 --field 10/445,200,545,250/signature10 --field 11/445,200,545,250/signature11 --field 12/445,200,545,250/signature12 --field 13/445,200,545,250/signature13 --field 14/445,200,545,250/signature14 --field 15/445,200,545,250/signature15 --field 16/445,200,545,250/signature16 --field 17/445,200,545,250/signature17 --field 18/445,200,545,250/signature18 --field 19/445,200,545,250/signature19 --field 20/445,200,545,250/signature20 --field 21/445,200,545,250/signature21 --field 22/445,200,545,250/signature22 --field 23/445,200,545,250/signature23 --field 24/445,200,545,250/signature24 --field 25/445,200,545,250/signature25 --field 26/445,200,545,250/signature26 --field 27/445,200,545,250/signature27 --field 28/445,200,545,250/signature28 --field 29/445,200,545,250/signature29 --field 30/445,200,545,250/signature30 --field 31/445,200,545,250/signature31  uploads/61d2b6c2f73ca51434986fdb0output.pdf uploads/61d2b6c2f73ca51434986fdb0_output.pdf`

I am using the above command to create 30 signature fields 

But when signing using `pyhanko sign addsig` how do i sign all the signature fields at once ?

*Note*- There is only one signer who will be signing.

**Comment by MatthiasValvekens** [comment, marked as answer]
Hi,

That's an excellent question. Unfortunately having a single signature with multiple visual appearances is something the PDF standard explicitly disallows (see ISO 32000-2:2020, 12.7.5.5, 3rd para). As long as that restriction is in place, it's unlikely that this will ever be supported in the CLI.

Here are your options:

 - Ignore the requirement and hack together a signature field with multiple appearances, then sign it using pyHanko. This is the hardest option because you'll really need to know your way around subclauses 12.5 and 12.7 of the standard; pyHanko won't help you all that much. Doing this might also result in PDF files that behave in unexpected ways.
 - If you know you'll only have a single "real" signer, you can use pyHanko in stamping mode (`pyhanko stamp`) to put a dummy signature appearance on every page, and then sign the entire document with an actual PDF signature.
 - Sign every field separately, as you are (probably) doing now. This will be relatively slow and produce bloated PDFs.

Regardless of which option you choose, you'll need to be aware of the following:

- Doing all of that in a single Python invocation is not currently possible through the pyHanko CLI. You'll need to wrap pyHanko in your own (minimal) Python library to pull it off. Having said that, in the case of option 3, you're probably not losing _that_ much to the interpreter's warm up time; serial PDF signing is inherently pretty slow.
- Consider the legal implications of your workflow. If your requirements are such that a signature indication must appear on every page for legal reasons in your jurisdiction, there's a chance that any signing workflow that does all of that in a single user interaction will not hold up to scrutiny in court. For what it's worth, this is (more or less) the reason why the spec restriction on multi-appearance signatures exists in the first place.

Hope that helps.

**Reply by santi2491** [reply]
Could you put a code example of how the signature would be stamped on several pages if we only have one signatory?

**Comment by ntqp97** [comment]
@MatthiasValvekens  Hi, I moved this discussion yesterday to this thread
Currently, I want to predefine two signature fields (signature_1 and signature_2) in the async_digest_doc_for_signing step.

def prepare_document(file_data, sig_name_1, sigpage_1, signature_box_1, signature_img_1,
                     sig_name_2, sigpage_2, signature_box_2, signature_img_2, cert, signer):
    ext_signer = MySignHelper.instantiate_external_signer(bytes(256), cert)
    
    def sign_pdf(pdf_data, sig_name, sigpage, signature_box, signature_img):
        pdf_signer = signers.PdfSigner(
            signature_meta=signers.PdfSignatureMetadata(
                field_name=sig_name,
                md_algorithm="sha256",
                location="Đà Nẵng",
                reason=f"{signer.name}<{signer.email}> đã ký lên văn bản này!",
                contact_info=f"{signer.email}",
                subfilter=SigSeedSubFilter.PADES
            ),
            new_field_spec=fields.SigFieldSpec(
                sig_field_name=sig_name,
                on_page=sigpage,
                box=signature_box,
            ),
            stamp_style=stamp.TextStampStyle(
                background=images.PdfImage(signature_img),
                stamp_text="",
                background_opacity=1,
                border_width=0
            ),
            signer=ext_signer,
        )
        prep_digest, tbs_document, output = asyncio.run(pdf_signer.async_digest_doc_for_signing(
            pdf_out=pdf_data,
            bytes_reserved=BYTES_RESERVED
        ))
        signed_attrs = asyncio.run(ext_signer.signed_attrs(
            data_digest=prep_digest.document_digest,
            digest_algorithm='sha256',
            use_pades=True,
        ))
        return prep_digest, tbs_document.post_sign_instructions, signed_attrs, output

    # Ký lần thứ nhất
    prep_digest_1, psi_1, signed_attrs_1, signed_output_1 = sign_pdf(
        file_data, sig_name_1, sigpage_1, signature_box_1, signature_img_1
    )

    # Ký lần thứ hai
    prep_digest_2, psi_2, signed_attrs_2, signed_output_2 = sign_pdf(
        signed_output_1, sig_name_2, sigpage_2, signature_box_2, signature_img_2
    )

    return prep_digest_1, psi_1, signed_attrs_1, prep_digest_2, psi_2, signed_attrs_2, signed_output_2
After that, I call the API for the third-party sign service and receive two signatures.
Finally, the signatures will be placed sequentially in these predefined fields

**Comment by ntqp97** [comment]
@MatthiasValvekens Is there any way I can do this myself? Could you suggest?
